Revise “Configure a Pod to Use a ConfigMap” task (#35750)
kubernetes/website: content/en/docs/tasks/configure-pod-container/configure-pod-configmap.md

---

<!-- overview -->
Many applications rely on configuration which is used during either application initialization
or runtime. Most of the times there is a requirement to adjust values assigned to configuration
parameters.
ConfigMaps are the Kubernetes way to inject application pods with configuration
data. ConfigMaps allow you to decouple configuration artifacts from image content to keep
containerized applications portable. This page provides a series of usage examples demonstrating

> how to create ConfigMaps and configure Pods using data stored in ConfigMaps.

Reviewer: This "line wraps" just reverted by your following commit.

```shell
kubectl create configmap <map-name> <data-source>
```

where \<map-name> is the name you want to assign to the ConfigMap and \<data-source> is the
directory, file, or literal value to draw the data from.
The name of a ConfigMap object must be a valid
[DNS subdomain name](/docs/concepts/overview/working-with-objects/names#dns-subdomain-names).

When you are creating a ConfigMap based on a file, the key in the \<data-source> defaults to
the basename of the file, and the value defaults to the file content.

You can use [`kubectl describe`](/docs/reference/generated/kubectl/kubectl-commands/#describe) or
[`kubectl get`](/docs/reference/generated/kubectl/kubectl-commands/#get) to retrieve information
about a ConfigMap.

#### Create a ConfigMap from a directory {#create-configmaps-from-directories}

You can use `kubectl create configmap` to create a ConfigMap from multiple files in the same
directory. When you are creating a ConfigMap based on a directory, kubectl identifies files
whose basename is a valid key in the directory and packages each of those files into the new
ConfigMap. Any directory entries except regular files are ignored (for example: subdirectories,

> symlinks, devices, pipes, and more).

Author: I amended the text - how's this?

> symlinks, devices, pipes, and more).

Reviewer 2: good.

<!-- overview -->
Many applications rely on configuration which is used during either application initialization or runtime.
Most times, there is a requirement to adjust values assigned to configuration parameters.
ConfigMaps are a Kubernetes mechanism that let you inject configuration data into application
{{< glossary_tooltip text="pods" term_id="pod" >}}.

The ConfigMap concept allow you to decouple configuration artifacts from image content to
keep containerized applications portable. For example, you can download and run the same
{{< glossary_tooltip text="container image" term_id="image" >}} to support local development,

> and for a system test, and in live end-user workload.

Author: Hmm. This new wording could make it sound like “local developmenr” is a kind of workload.

I agree it could be even better than I've proposed.

> {{< glossary_tooltip text="container image" term_id="image" >}} to support local development,
> and for a system test, and in live end-user workload.

Reviewer 3: How about? 
```suggestion
{{< glossary_tooltip text="container image" term_id="image" >}} to spin up containers for the purposes of local development, system test, or running a live end-user workload.
```

configure Pods using data stored in ConfigMaps.

## {{% heading "prerequisites" %}}

{{< include "task-tutorial-prereqs.md" >}}

You need to have the `wget` tool installed. If you have a different tool
such as `curl`, and you do not have `wget`, you will need to adapt the
step that downloads example data.

<!-- steps -->

## Create a ConfigMap

You can use either `kubectl create configmap` or a ConfigMap generator in `kustomization.yaml`
to create a ConfigMap.

### Create a ConfigMap using `kubectl create configmap`

Use the `kubectl create configmap` command to create ConfigMaps from
[directories](#create-configmaps-from-directories), [files](#create-configmaps-from-files),
or [literal values](#create-configmaps-from-literal-values):

```shell
kubectl create configmap <map-name> <data-source>
```

where \<map-name> is the name you want to assign to the ConfigMap and \<data-source> is the
directory, file, or literal value to draw the data from.
The name of a ConfigMap object must be a valid
[DNS subdomain name](/docs/concepts/overview/working-with-objects/names#dns-subdomain-names).

When you are creating a ConfigMap based on a file, the key in the \<data-source> defaults to

> the basename of the file, and the value defaults to the file content.

Author: What should that note say?

> the basename of the file, and the value defaults to the file content.

Reviewer 2: Maybe something like:

```
When the provided file is an ASCII file, the basename of the file is stored into `binaryData`
field of the ConfigMap, with the base64-encoded file content as its value.
Also, the file size is subject to the ConfigMap constraint which means the file cannot be
larger than 1 MB.
```

       --from-env-file=configure-pod-container/configmap/game-env-file.properties
```

would produce a ConfigMap. View the ConfigMap:

```shell
kubectl get configmap game-config-env-file -o yaml
```

the output is similar to:
```yaml
apiVersion: v1
kind: ConfigMap
metadata:

>   creationTimestamp: 2019-12-27T18:36:28Z

Reviewer: Not sure why use 2019 instead of 2017, while line 133 use 2022.

Author: I reckon it'd be fine to bump that `creationTimestamp`.

Author: Unless we _have_ to bump this, let's leave it.

apiVersion: v1
kind: ConfigMap
metadata:

>   creationTimestamp: 2022-02-18T18:54:22Z

Reviewer 3: Maybe I'm missing something here, but are we bumping the dates based on some specific requirement and according to a specific pattern?

Author: Not really, I just like our docs to look maintained.

Reviewer 3: Ah okay, thanks for clarifying!

Apply the kustomization directory to create the ConfigMap object.
```shell
kubectl apply -k .
```
```
configmap/game-config-5-m67dt67794 created
```

#### Generate ConfigMaps from literals

This example shows you how to create a `ConfigMap` from two literal key/value pairs:
`special.type=charm` and `special.how=very`, using Kustomize and kubectl. To achieve
this, you can specify the `ConfigMap` generator. Create (or replace)
`kustomization.yaml` so that it has the following contents:

```yaml
---
# kustomization.yaml contents for creating a ConfigMap from literals
configMapGenerator:
- name: special-config-2
  literals:
  - special.how=very
  - special.type=charm
```

Apply the kustomization directory to create the ConfigMap object:
```shell
kubectl apply -k .
```
```
configmap/special-config-2-c92b5mmcf2 created
```

Now that you have learned to define ConfigMaps, you can move on to the next
section, and learn how to use these objects with Pods.

> 

Author: I added some cleanup steps.

volume will be empty. If the ConfigMap exists, but the referenced key is non-existent the path
will be absent beneath the mount point. See [Optional ConfigMaps](#optional-configmaps) for more
details.

### Mounted ConfigMaps are updated automatically

When a mounted ConfigMap is updated, the projected content is eventually updated too.
This applies in the case where an optionally referenced ConfigMap comes into
existence after a pod has started.

Kubelet checks whether the mounted ConfigMap is fresh on every periodic sync. However,
it uses its local TTL-based cache for getting the current value of the ConfigMap. As a
result, the total delay from the moment when the ConfigMap is updated to the moment
when new keys are projected to the pod can be as long as kubelet sync period (1
minute by default) + TTL of ConfigMaps cache (1 minute by default) in kubelet. You
can trigger an immediate refresh by updating one of the pod's annotations.

{{< note >}}
A container using a ConfigMap as a [subPath](/docs/concepts/storage/volumes/#using-subpath) volume will not receive ConfigMap updates.
{{< /note >}}

<!-- discussion -->

> ## Understanding ConfigMaps and Pods

Author: Are you saying that this should happen in this PR @tengqm? I'm worried about this PR becoming larger than it needs to be.

> ## Understanding ConfigMaps and Pods

Reviewer 2: It doesn't have to be.

The ConfigMap API resource stores configuration data as key-value pairs. The data can be consumed
in pods or provide the configurations for system components such as controllers. ConfigMap is
similar to [Secrets](/docs/concepts/configuration/secret/), but provides a means of working
with strings that don't contain sensitive information. Users and system components alike can
store configuration data in ConfigMap.

{{< note >}}
ConfigMaps should reference properties files, not replace them. Think of the ConfigMap as
representing something similar to the Linux `/etc` directory and its contents. For example,
if you create a [Kubernetes Volume](/docs/concepts/storage/volumes/) from a ConfigMap, each
data item in the ConfigMap is represented by an individual file in the volume.
{{< /note >}}

The ConfigMap's `data` field contains the configuration data. As shown in the example below,

> this can be simple -- like individual properties defined using `--from-literal` -- or complex --
> like configuration files or JSON blobs defined using `--from-file`.

Reviewer 4: ```suggestion
this can be simple (like individual properties defined using `--from-literal`) or complex
(like configuration files or JSON blobs defined using `--from-file`).
```
This is a little easier for my brain to parse, but I leave that to you whether or not you find it helpful.

    property.3=value-3
```

When `kubectl` creates a ConfigMap from inputs that are not ASCII or UTF-8, the tool puts
these into the `binaryData` field of the ConfigMap, and not in `data`. Both text and binary
data sources can be combined in one ConfigMap.

If you want to view the `binaryData` keys (and their values) in a ConfigMap, you can run
`kubectl get configmap -o jsonpath='{.binaryData}' <name>`.

Pods can load data from a ConfigMap that uses either `data` or `binaryData`.

## Optional ConfigMaps

You can mark a reference to a ConfigMap as _optional_ in a Pod specification.
If the ConfigMap doesn't exist, the configuration for which it provides data in the Pod

> (e.g. environment variable, mounted volume) will be empty.

Reviewer 4: ```suggestion
(e.g. an environment variable or a mounted volume) will be empty.
```
